### PR TITLE
http2: client round 1

### PR DESCRIPTION
**Pull request by mcollina**
@jasnell this is not working, I suspect it is not sending anything on the wire in fact.

### CODE REVIEW

File: lib/internal/http2/core.js [hunk 1]
```diff
@@ -757,6 +771,15 @@ class ServerHttp2Stream extends Http2Stream {
   }
 }
 
+ServerHttp2Stream.prototype._proceed = ServerHttp2Stream.prototype.respond;
```
Comment (jasnell) [inline lib/internal/http2/core.js]: Let's use a Symbol instead of `_proceed`

**Review by jasnell — COMMENTED** [review]
Quick read through... most bits look in order. Will pull it down and test in just a bit tho

**Comment by mcollina** [comment]
@jasnell updated with `kProceed`.

**Comment by jasnell** [comment]
+1... I started digging in yesterday then completely ran out of time to finish. Hoping to return to it this afternoon.

File: lib/internal/http2/core.js [hunk 2]
```diff
@@ -571,7 +586,7 @@ class Http2Stream extends Duplex {
     unrefTimer(this);
     var state = this[kState];
     if (!state.headersSent)
-      this.respond();
+      this._proceed();
```
Comment (jasnell) [inline lib/internal/http2/core.js]: `s/_proceed/[kProceed]` ??

Comment (mcollina) [reply lib/internal/http2/core.js]: Ouch, yes. Done.

**Comment by jasnell** [comment]
Ok, looks like there's a conflict with the change in node_http2.cc that needs to be looked at following @addaleax's refactor that landed yesterday. I'll look at resolving that in just a bit

**Comment by mcollina** [comment]
@jasnell this is not working ATM. You can check out the branch as it is and let me know what the problem is.

This is my current client, and it's not working :(:

```js
'use strict'

const net = require('net');
const h2 = require('http2');

const socket = net.connect(8000, 'localhost');
const client = h2.createClientSession(socket);

const req = client.request();

socket.on('connect', function () {
  console.log('socket connected');
});

req.on('response', function (headers) {
  console.log('response!', headers)
})

let data = '';
req.setEncoding('utf8');
req.on('data', console.log);
```

**Comment by jasnell** [comment]
Yeah, I've applied the patch locally and will be going through it in detail again here shortly. Hoping to have it working by the time we meet up in Seattle

**Comment by jasnell** [comment]
Here's the "working" client 

```js
'use strict'

const common = require('../common');
const assert = require('assert');
const h2 = require('http2');
const net = require('net');
const body =
  '<html><head></head><body><h1>this is some data</h2></body></html>';

const server = h2.createServer();

server.on('stream', common.mustCall(onStream));

function onStream (stream) {
  stream.respond({
    'content-type': 'text/html',
    ':status': 200
  })
  stream.end(body);
}

server.listen(0);

server.on('listening', common.mustCall(function() {
  const socket = net.connect(this.address());
  const client = h2.createClientSession(socket);

  setImmediate(() => {
    const headers = {
      ':method': 'GET',
      ':scheme': 'http',
      ':authority': `localhost:${this.address().port}`,
      ':path': '/'
    };

    const req = client.request(headers);

    req.on('response', common.mustCall(function (headers) {
      console.log('response!', headers)
    }))

    let data = '';
    req.setEncoding('utf8');
    req.on('data', d => data += d);
    req.on('end', common.mustCall(function () {
      assert.strictEqual(body, data);
    }));
    req.end();
  });

}));
```

Also, in core.js, `createClientSession`, send the client settings in setImmediate: 

```js
  setImmediate(() => session.submitSettings(settings));
```

**Comment by mcollina** [comment]
that's progress!
I've updated the test and it is _almost_ passing, but it's giving me a:

```
Assertion failed: (handle->ready_callbacks_head == nullptr), function nghttp2_session_free, file ../src/node-http2-core-inl.h, line 845.
Abort trap: 6
```

I'll work on it later on.

**Comment by mcollina** [comment]
@jasnell this is ready to go. The test is green with a bunch of TODO!

**Comment by jasnell** [comment]
Landed!

**Pull request closed without merging**